common/scala/src/main/scala/org/apache/openwhisk/core/database/ActivationStore.scala

```diff
@@ -35,21 +35,25 @@ case class UserContext(user: Identity, request: HttpRequest = HttpRequest())
 trait ActivationStore {
 
   protected val disableStoreResultConfig = loadConfigOrThrow[Boolean](ConfigKeys.disableStoreResult)
+  protected val disableStoreNonBlockingResultConfig = loadConfigOrThrow[Boolean](ConfigKeys.disableStoreNonBlockingResult)
   protected val unstoredLogsEnabledConfig = loadConfigOrThrow[Boolean](ConfigKeys.unstoredLogsEnabled)
 
   /**
    * Checks if an activation should be stored in database and stores it.
    *
    * @param activation activation to store
    * @param isBlockingActivation is activation blocking
+   * @param disableBlockingStore do not store activation if successful and blocking
+   * @param disableNonBlockingStore do not store activation if successful and non-blocking
```

Comment:
Widening the signature here this way strikes me as confusing - couldn't this be better done with an enum: `always store`, `never store`, `store for non blocking only`, etc?

Reply:
I guess my question would be how could you do that without breaking the config that already exists for `disableBlockingStore` that's presumably already being used?

The other thing is I think people still want to distinguish between successful and failures so they only want to turn off writes for successes. I'm not sure what the best solution is or what options should be offered.

I think I understand your suggestion in that there could be an enum set with values `ALWAYS_STORE_SUCCESSES`, `STORE_BLOCKING_SUCCESSES`, `STORE_NONBLOCKING_SUCCESSES`, `STORE_NO_SUCCESSES`. However I think doing this would break deployments unless we kept it as two booleans in the config as currently is in the pr and built the enum from the boolean values. But I think that defeats the purpose of setting up the enum at all.

And then on our discussion for our use case specifically with only needing specific failure types written (developer and system). I think that should be a separate check which could either be an enum or just a boolean of `storeApplication Errors` since I think developer and system errors both are very similar in what they mean. The divider in failures to me is really just whether it's a user application error or not.

Reply:
I think there are two orthogonal issues - one is configuration and making that backward compatible. The other is the internal interface and implementation. 

I think the distinctions you want to make suggests this alternate/simpler enum:

- `STORE_ALWAYS` (default)
- `STORE_ON_FAILURE`
- `STORE_NON_BLOCKING`

From a configuration point of view, you can continue to accept the current configuration property (perhaps with a deprecation warning) and introduce a new configuration that maps to the enum. It would be an error to specify both configurations properties.

Reply:
I agree with that setup. The one case that I think is confusing is the existing boolean and wanting failed blocking activations to still be stored. Because setting the current boolean means store failed blocking activations and non-blocking activations. So that falls between `STORE_ON_FAILURE` and `STORE_NON_BLOCKING`

Reply:
I was thinking that `STORE_ON_FAILURE` applies to everything. I suppose we can have two modes for this: `STORE_ON_FAILURE_BLOCKING_AND_NON_BLOCKING` for sync/async and `STORE_ON_FAILURE_NON_BLOCKING_ONLY` (and so OK there are four again :)

Reply:
This is what I have. I can't think of a scenario where someone would want a subset of non blocking activations written, but not a larger subset for blocking. I can only think of wanting a subset of blocking and wanting a larger subset for non-blocking. I'll update the pr if you're good with this

`STORE_ALWAYS`
`STORE_NON_BLOCKING_OR_ON_BLOCKING_FAILURE`
`STORE_ON_FAILURE`
`STORE_ON_NON_APPLICATION_FAILURE`

Reply:
`STORE_NON_BLOCKING_OR_ON_BLOCKING_FAILURE` and `STORE_ON_FAILURE` are redundant for blocking invokes, no?

Reply:
Yea so what I was originally going to say as well is that I think we're just getting close to having a separate config for each invocation type, blocking and non-blocking. In your list, the original person that created the blocking config wanted only blocking failures and all non-blocking; which is why I had `STORE_NON_BLOCKING_OR_ON_BLOCKING_FAILURE`. Almost everyone would want this because not storing any non-blocking means you can't run sequences I believe. But you are correct that would mean the parameters aren't orthogonal, which is why I think we're at the point of just having two separate params with the same enum.

`STORE_ALWAYS`
`STORE_FAILURES`
`STORE_FAILURES_NOT_APPLICATION_ERRORS`